File: backend/app/schemas/prompt.py
from pydantic import BaseModel, Field
from typing import List, Optional
from datetime import datetime
from app.models.prompt import PromptStatus


# Base schemas
class PromptBase(BaseModel):
    """Base prompt schema - simplified"""
    title: str
    description: str
    content: str
    category_id: str


# Request schemas
class PromptCreate(PromptBase):
    """Prompt creation schema"""
    image_url: Optional[str] = None
    is_featured: bool = False


class PromptUpdate(BaseModel):
    """Prompt update schema"""
    title: Optional[str] = None
    description: Optional[str] = None
    content: Optional[str] = None
    image_url: Optional[str] = None
    category_id: Optional[str] = None
    status: Optional[PromptStatus] = None
    is_featured: Optional[bool] = None


class PromptFilter(BaseModel):
    """Prompt filtering schema"""
    category_id: Optional[str] = None
    status: Optional[PromptStatus] = None
    is_featured: Optional[bool] = None
    search: Optional[str] = None


class PromptSort(BaseModel):
    """Prompt sorting schema"""
    field: str = "created_at"
    order: str = "desc"  # asc or desc


# Response schemas
class PromptResponse(PromptBase):
    """Prompt response schema"""
    id: str
    image_url: Optional[str] = None
    status: PromptStatus
    is_featured: bool
    is_active: bool
    likes_count: int
    created_at: datetime
    updated_at: datetime
    
    model_config = {"from_attributes": True}


class PromptSummary(BaseModel):
    """Prompt summary schema for lists"""
    id: str
    title: str
    description: str

    image_url: Optional[str] = None
    category_id: str
    is_featured: bool
    likes_count: int
    created_at: datetime
    
    model_config = {"from_attributes": True}


class PromptDetail(PromptResponse):
    """Detailed prompt schema"""
    category_name: Optional[str] = None
    is_favorited: bool = False
    related_prompts: List[PromptSummary] = []


class PromptStats(BaseModel):
    """Prompt statistics schema"""
    total_prompts: int
    published_prompts: int
    draft_prompts: int
    featured_prompts: int
    premium_prompts: int
    total_views: int
    total_likes: int
    average_rating: float


# Admin schemas
class PromptAdmin(PromptResponse):
    """Admin prompt schema with additional fields"""
    created_by: Optional[str] = None

File: admin/src/types/prompt.ts
/**
 * Prompt related types
 */

import { BaseEntity, PaginationParams, SearchParams, SortParams } from './common';

export type PromptStatus = 'draft' | 'published';

export interface BasePrompt {
  title: string;
  description: string;
  content: string;
  category_id: string;
}

export interface Prompt extends BaseEntity, BasePrompt {
  image_url?: string;
  status: PromptStatus;
  is_featured: boolean;
  is_active: boolean;
  likes_count: number;
  created_by?: string;
}

export interface PromptCreate extends BasePrompt {
  image_url?: string;
  is_featured?: boolean;
  status?: PromptStatus;
}

export interface PromptUpdate {
  title?: string;
  description?: string;
  content?: string;
  image_url?: string;
  category_id?: string;
  status?: PromptStatus;
  is_featured?: boolean;
  is_active?: boolean;
}

export interface PromptSummary {
  id: string;
  title: string;
  description: string;
  image_url?: string;
  category_id: string;
  is_featured: boolean;
  likes_count: number;
  created_at: string;
}

export interface PromptDetail extends Prompt {
  category_name?: string;
}

export interface PromptAdmin extends Prompt {
  // Inherits all fields from Prompt - no additional fields needed
}

export interface PromptFilters extends PaginationParams, SearchParams, SortParams {
  status?: PromptStatus;
  category_id?: string;
  is_featured?: boolean;
}

export interface PromptStats {
  total_prompts: number;
  published_prompts: number;
  draft_prompts: number;
  featured_prompts: number;
  total_likes: number;
}

export interface PromptChartData {
  prompts_added: number[];
  trending_prompts: number[];
  totals: {
    total_prompts_added: number;
    total_trending: number;
  };
}
